include/moveit/drake/conversions.hpp

```diff
@@ -0,0 +1,116 @@
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Sebastian Jahr
+   Desc: TODO
+*/
+
+#include <moveit/planning_scene/planning_scene.h>
+#include <moveit/robot_trajectory/robot_trajectory.h>
+
+#include "drake/multibody/parsing/parser.h"
+#include "drake/geometry/scene_graph.h"
+#include "drake/systems/framework/diagram.h"
+#include "drake/systems/framework/diagram_builder.h"
+#include "drake/multibody/plant/multibody_plant.h"
+#include <drake/common/trajectories/trajectory.h>
+#include <drake/common/trajectories/piecewise_polynomial.h>
+namespace moveit::drake
+{
+
+using ::drake::geometry::SceneGraph;
+using ::drake::multibody::AddMultibodyPlantSceneGraph;
+using ::drake::multibody::MultibodyPlant;
+using ::drake::multibody::Parser;
+using ::drake::systems::DiagramBuilder;
+
+/**
+ * @brief Create a Piecewise Polynomial from a moveit trajectory (see
+ * https://drake.mit.edu/doxygen_cxx/classdrake_1_1trajectories_1_1_piecewise_polynomial.html)
+ *
+ * @param robot_trajectory MoveIt trajectory to be translated
+ * @param group Joint group for which a piecewise polynomial is created
+ * @return ::drake::trajectories::PiecewisePolynomial<double>
+ */
+[[nodiscard]] ::drake::trajectories::PiecewisePolynomial<double>
+getPiecewisePolynomial(const ::robot_trajectory::RobotTrajectory& robot_trajectory,
+                       const moveit::core::JointModelGroup* group)
+{
+  std::vector<double> breaks;
+  std::vector<Eigen::MatrixXd> samples;
+
+  // Create samples & breaks
+  for (std::size_t i = 0; i < robot_trajectory.getWayPointCount(); ++i)
+  {
+    const auto& state = robot_trajectory.getWayPoint(i);
+    Eigen::VectorXd position(state.getVariableCount());
+    state.copyJointGroupPositions(group, position);
+    samples.emplace_back(position);
+    breaks.emplace_back(robot_trajectory.getWayPointDurationFromStart(i));
+  }
+
+  // Create a piecewise polynomial trajectory
+  return ::drake::trajectories::PiecewisePolynomial<double>::FirstOrderHold(breaks, samples);
+}
+
+/**
+ * @brief Create a moveit trajectory from a piecewise polynomial. Assumes that the piecewise polynomial describes a
+ * joint trajectory for every active joint of the given trajectory.
+ *
+ * @param piecewise_polynomial Drake trajectory
+ * @param samples Number of timepoints at which a waypoint is created
+ * @param output_trajectory MoveIt trajectory to be populated based on the piecewise polynomial
+ */
+void getRobotTrajectory(const ::drake::trajectories::PiecewisePolynomial<double>& piecewise_polynomial,
+                        const int samples, std::shared_ptr<::robot_trajectory::RobotTrajectory>& output_trajectory)
+{
+  // Get the start and end times of the piecewise polynomial
+  const auto time_step = piecewise_polynomial.end_time() / static_cast<double>(samples - 1);
+
+  for (double t = 0.0; t <= piecewise_polynomial.end_time(); t += time_step)
```

Comment:
I think this might be the issue... you're not exactly hitting the last time step.

You should instead consider using an integer iteration variable and then saturating to the real end time of the trajectory. See what I did in the latest PR for KTOpt.